### DISCUSSION TITLE
How to insert custom Vue Node in TipTap editor field?

### DISCUSSION THREAD

**Discussion by Arturexe** [ueberdosis/tiptap#2849] [General Questions]
I want to insert a `vue component` into the editor field. At this point I'm not even sure if this has to be a `node`, an `extension` or a `mark`. The [docs][1] explain it with `nodes` so I assume that that's it.

So I've created this `Component.vue`:

    <template>
      <node-view-wrapper class="vue-component">
        Vue Component
        <node-view-content as="p"/>
      </node-view-wrapper>
    </template>
    
    
    <script setup>
    import { NodeViewWrapper, NodeViewContent } from '@tiptap/vue-3'
    
    
    const props = defineProps({
        node: {
            type: Object,
            required: true
        },
        updateAttributes: {
            type: Function,
            required: true,
        }
    })
    </script>

and integrated it in a new `node-extension.js`:

    import { Node } from '@tiptap/core'
    import { VueNodeViewRenderer } from '@tiptap/vue-3'
    import Component from './Component.vue'
    
    export default Node.create({
        // configuration …
    
        addNodeView() {
            return VueNodeViewRenderer(Component)
        }
    })

but how do I insert it in the `editor field`? I've imported it in the `editor component`:

    import CustomNode from '../../../scripts/node-extension.js'

and added it to the `editor object`:

    const editor = useEditor({
        extensions: [
            StarterKit,
            TextAlign.configure({ types: ['heading', 'paragraph'] }),
            Underline,
            Link,
            Image,
            CustomNode
        ],
        content: props.modelValue,
        onUpdate: ({ editor }) => {
            emit('update:modelValue', editor.getHTML())
        },
    })

but what is the command to insert `content`/`node` into the editor field? I've tried:

    editor.chain().focus().VueNodeViewRenderer(CustomNode).run()
    editor.chain().focus().addNodeView(CustomNode).run()
    editor.chain().focus().insertContent(CustomNode).run()

but its either nothing happening or `this is not a function` error being thrown. I really wish the [docs][1] would be more extensive about this.

  [1]: https://tiptap.dev/api/nodes

**Comment by sereneinserenade** [comment]
You have to have a command in your custom extension, which will set your custom node.
```
// node-extension.js
import { Node } from '@tiptap/core'
import { VueNodeViewRenderer } from '@tiptap/vue-3'
import Component from './Component.vue'

export default Node.create({
    // configuration …

    addNodeView() {
        return VueNodeViewRenderer(Component)
    },

    addCommands() {
      return {
         yourCommand: () => ({ commands }) => commands.insertContent("<yourNode> something </yourNode>")
      }
    }
})
```

and then you run it via `editor.chain().focus().yourCommand().run()`

### example:

take a look here https://github.com/ueberdosis/tiptap/blob/main/packages/extension-horizontal-rule/src/horizontal-rule.ts#L46 you can run it with `editor.chain().focus().setHorizontalRule().run()`
